datapreprocessing.py

```diff
@@ -0,0 +1,43 @@
+import pandas as pd
+import pytz
+import re
+from libs.helpermodules import memory_handling as mh
+
+# Load the data
+df = pd.read_csv('fedspeeches.csv')
+
+# Convert the date column to datetime
+df['date'] = pd.to_datetime(df['date'])
+
+# Define the timezone for EST
+est = pytz.timezone('US/Eastern')
+
+# Add a timestamp column (assuming speeches start at 10:00 AM EST)
+df['timestamp'] = df['date'].apply(lambda x: est.localize(x.replace(hour=10, minute=0, second=0)))
+
+# Estimate the length of the speech in minutes (assuming average speaking rate of 130 words per minute)
+df['speech_length_minutes'] = df['text'].apply(lambda x: max(1, len(x.split()) / 130))
+
+
+def split_text_by_minute(text, minutes):
+    words = text.split()  # Split by whitespace
+    words_per_minute = max(1, len(words) // minutes)
+    return [' '.join(words[i:i + words_per_minute]) for i in range(0, len(words), words_per_minute)]
+
+
+# Apply the function to split the text
+df['text_by_minute'] = df.apply(lambda row: split_text_by_minute(row['text'], int(row['speech_length_minutes'])),
+                                axis=1)
+
+# Expand the dataframe
+df_expanded = df.explode('text_by_minute').reset_index(drop=True)
+
+# Add minute offset to the timestamp
+df_expanded['minute'] = df_expanded.groupby('timestamp').cumcount()
+df_expanded['timestamp'] = df_expanded['timestamp'] + pd.to_timedelta(df_expanded['minute'], unit='m')
+
+# Drop unnecessary columns
+df_expanded = df_expanded.drop(columns=['minute', 'speech_length_minutes'])
+
+pickle_helper = mh.PickleHelper(df_expanded)
+pickle_helper.pickle_dump('fedspeeches_preprocessed')
```

Comment:
send me a picture of what you get in the pickle when you process it

Reply:
I am not 100% sure about this, but i think the .gitmodules file is needed to make the helpermodules submodule work (otherwise if other people clone this repo it does  not get downloaded automatically). I can make those changes to datapreprocessing if needed, but right now it is not really meant to be a "helper module", (because it just does one very specific thing which i am not sure would ever be ever again). But I can definitely turn it into one, I just need to know which parts are the ones that you'd like to keep (like the text cleaning or the speech subdivision). As per the last request, I am not 100% sure what you mean, but this is a sample of the pkl binary file:
![Screenshot 2024-11-17 122516](https://github.com/user-attachments/assets/fc052675-40b4-43bf-a4eb-bd83ad5b8489)
![Screenshot 2024-11-17 122554](https://github.com/user-attachments/assets/1a0459b0-314e-4ce3-ba99-04dfa7d1e4c4)

Reply:
The parts you need to keep are the ones you need for the project, even if the function is very specific it's still better to have the repository as clean as possible. Look at [portfolioreturns_visualizer](https://github.com/quantfin-sfpolimi/portfolioreturns_visualizer) for example. In it there are no functions but only the main (with as little code as possible) and the data needed, so even if the functions will be useless after this project (we don't know for sure) it's better if you put it in helpermodules anyway. 

Thanks for the screenshots, about .gitmodules ask Giulia. I did the review with her and I am still learning so I'm not 100% sure about the working of GitHub that's why I would like you (and the others) to be as critical as possible of my comments so that I can learn faster.

I hope to have provided a satisfactory answer, in case you need anything don't esitate to contact me.

Reply:
I looked at the screenshot. The first image is of df_expanded ? In case it is the data should be cleaned, I don't know if you intend to do it now or later but if the model you intend to use where to interpretate those characters as a Cb communications it will surely be problematic.

Reply:
you shouldn’t include .gitmodules in your push request unless it’s intentional. It can mess up the submodule settings for others, break pipelines, ecc. in your case, it wouldnt have broken anything, but it has happened in the past, which is why for now we avoid it

This is Giulia's answer about .gitmodules. Basically it's just risky

Reply:
Ok, so I am going to put .gitmodules in my .gitignore folder. As per the pickle file, the df is clean, I believe that's just the way pickle files compress the information. If you want to visualize it in a more "human-friendly" way I can send you a screenshot of the dataframe opened in python. I am also going to put the other function in helpermodules.